Docs for fast deploys (#10876)
dagster-io/dagster: docs/content/dagster-cloud/deployment/serverless.mdx

> Finally, deploy your project with Dagster Cloud Serverless:

Reviewer: do we want to include instructions for both the pex path and the non pex path?

```shell

> dagster-cloud serverless deploy-python-executable \

Reviewer: bit unfortunate that the new default is the one with the non-default name (i.e. should `deploy` now be `deploy-image` or something?)

Author: I didn't want to break existing users. Going forward we can do this:
1. Add `deploy-image` that does docker deploy and tell users to migrate.
2. At some later point, change `deploy` to do `deploy-python-executable` instead of `deploy-image`
3. Later, remove `deploy-image`

---

## Customizing the runtime environment

> Dagster Cloud Serverless packages your code as [PEX](https://pex.readthedocs.io) files and deploys them on Docker images. Many apps will work fine with the default Dagster Cloud Serverless setup. However, some apps may need to make changes to the runtime environment, either to include data files, use a different OS, different Python version, or install some native dependencies. You can customize the runtime environment using various methods described below.

Reviewer: Should we explain why we do this and what benefits it provides?

### Including data files

> To add data files to your deployment, use the [Data Files Support](https://setuptools.pypa.io/en/latest/userguide/datafiles.html) built into Python's setup.py. This requires adding a `package_data` or `include_package_data` keyword in the call to `setup()` in `setup.py`.

Reviewer: linking to or showing an example here would be pretty helpful i think

Reviewer 2: +1

### Using a different Python version

> The default version of Python for Serverless deployments is Python 3.8. Versions 3.7, 3.9 and 3.10 are also supported. You can specify the version you want by updating your GitHub workflow or using the `--python-version` command line argument:

Reviewer: any particular reason the default isn't the latest version?

Author: No reason. It's just the one we use. Should I change it to 3.10?

Reviewer 3: Probably the most compelling reason is that all of our internal tests run using 3.8.

> - **With GitHub**: Change the `python_version` specified in your `.github/workflows` files. For example, change [this line](https://github.com/dagster-io/quickstart-etl/blob/e07e944c7504a52b3d252553d51ad2085b4d5914/.github/workflows/deploy.yml#L44) in your `deploy.yml` to the version you want.

Author: Done

- **With the CLI**: Add the `--python-version` CLI argument to the deploy command to specify the registry path to the desired base image:

  ```shell
  dagster-cloud serverless deploy --location-name=my_location --python-version=3.9
  ```

> ### Using a different OS or using native dependencies

Reviewer: s/different OS/different base image?

Your code is run on a Docker image built using:
1. The standard Python "slim" [Docker image](https://hub.docker.com/_/python), such as `python:3.8-slim`.

> 2. The `dagster-cloud[serverless]` module installed in the image.

Reviewer 2: Is this describing what we do? Or something the user does / needs to do on their end? Apologies if this is obvious, but as someone who's unfamiliar, I read this and I'm not sure what to do with it.

Author: This describes what we do. We could make that clear "We run your code on a Docker image that we build using:"

> As far as possible, add all dependencies by including the corresponding native Python bindings in your `setup.py`.   When that is not possible, you can build and upload a custom base image that will be used to run your Python code. 

Reviewer: nit extra spaces

Reviewer 2: Not a nit! +1

To build and upload the image, use the command line:

1. Build your docker image. Ensure you have included the `dagster-cloud[serverless]` dependency. You can do this by adding the following to your `Dockerfile`: `RUN pip install "dagster-cloud[serverless]"`.

> 2. Get the credentials to access the Dagster cloud image registry. The following should print out the registry URL:

Reviewer 2: ```suggestion
2. Get the credentials to access the Dagster Cloud image registry. The following should print out the registry URL:
```

```shell
$ set -a
$ eval "$(dagster-cloud serverless registry-info)"
$ set +a
$ echo $REGISTRY_URL
657821118200.dkr.ecr.us-west-2.amazonaws.com/serverless-agent-UUID
```

> 3. Login to docker using the retrieved credentials:

Reviewer 2: ```suggestion
3. Log in to Docker using the retrieved credentials:
```

> ```shell
> $ echo $AWS_ECR_PASSWORD | docker login --username AWS --password-stdin $REGISTRY_URL
> Login Succeeded
> ```

Reviewer 2: Same here, re: indenting the code block:

```suggestion
   ```shell
   $ echo $AWS_ECR_PASSWORD | docker login --username AWS --password-stdin 
   $REGISTRY_URL
   Login Succeeded
   ```
```

3. Tag your built docker image using the `$REGISTRY_URL`:

```shell

> $ docker tag build-image:current-tag $REGISTRY_URL:published-tag

Reviewer: I had sort of been imagining that this would be exposed by some CLI rather than having users interact directly with our ECR repo (I know this is what the github action does under the hood, but putting it behind a CLI gives us a bit more flexibility to adjust it in the future I think?)

added @Ramshackle-Jamathon and @prha in case they have thoughts  here

Reviewer 4: yeah, I like having this layer of indirection so that we can change it under the hood if we need to.

Author: Agreed. I'm working on a command to do this.

Author: What about this flow, before I implement it:

1. build your docker image as `your-image:your-tag`
2. `dagster-cloud serverless deploy-base-image your-image:your-tag published-tag`
3. Options:
    * (CLI) `dagster-cloud serverless deploy-python-executable ... --base-image-tag=published-tag`
    * (Github) `SERVERLESS_BASE_IMAGE_TAG: "published-tag"`

Author: This is now being changed in an internal PR. I'll revise the docs shortly.

Author: Done.

```

> 4. Publish your image:

Reviewer 2: This should be `5.`

> ```shell
> $ docker push $REGISTRY_URL:published-tag
> ```

Reviewer 2: Indenting this so it's under the relevant instructions:

```suggestion
   ```shell
   $ docker push $REGISTRY_URL:published-tag
   ```
```

> To use a docker image you have published to Dagster cloud:

Reviewer 2: `docker` > `Docker` ?

- **With GitHub**: Set the `SERVERLESS_BASE_IMAGE` environment variable your GitHub Actions configuration (usually at `.github/workflows/deploy.yml`):

  ```yaml

> 

Reviewer 2: Can you remove the extra space here?

  env:
    DAGSTER_CLOUD_URL: ...
    DAGSTER_CLOUD_API_TOKEN: ...

>     SERVERLESS_BASE_IMAGE: "657821118200.dkr.ecr.us-west-2.amazonaws.com/serverless-agent-UUID:published-tag"

Reviewer: I'm also surprised to see us referencing a specific AWS account here and ECR registry here rather than just a tag

```
- setup.py
- my_dagster_project/
  - __init__.py
  - repository.py
  - data/
    - file1.txt
    - file2.csv
```

If you want to include the `data` folder, modify your `setup.py` to add the `package_data` line:

```python
# setup.py
from setuptools import find_packages, setup

if __name__ == "__main__":
    setup(
        name="my_dagster_project",
        packages=find_packages(exclude=["my_dagster_project_tests"]),
        # Add the following line. Here "data/*" is relative to the my_dagster_project sub directory.
        package_data={"my_dagster_project": ["data/*"]},
        install_requires=[
            "dagster",
            ...
        ],
    )
```

### Using a different Python version

The default version of Python for Serverless deployments is Python 3.8. Versions 3.7, 3.9 and 3.10 are also supported. You can specify the version you want by updating your GitHub workflow or using the `--python-version` command line argument:

- **With GitHub**: Change the `python_version` parameter for the `build_deploy_python_executable` job in your `.github/workflows` files. For example:

> ```yaml

Reviewer 2: This should be indented to the same level as its content.
